src/event/discriminant.rs

```diff
@@ -108,6 +113,20 @@ fn hash_f64<H: Hasher>(hasher: &mut H, value: &f64) {
     hasher.write(&value.to_ne_bytes());
 }
 
+fn hash_array<H: Hasher>(hasher: &mut H, array: &Vec<Value>) {
+    for val in array.iter() {
+        hash_value(hasher, val);
+    }
+}
+
+fn hash_map<H: Hasher>(hasher: &mut H, map: &HashMap<Atom, Value>) {
+    // FIXME: until we replaced HashMap by IndexMap this might be broken
```

Comment:
This comment is on point - it'll most likely be broken. Consider implementing a slower workaround for now - or just switching to `IndexMap` right away.

Reply:
Just wanted to point out it's `BTreeMap` that solves this problem, not `IndexMap` :smile: 

`BTreeMap` is sorted by key. `IndexMap` maintains insertion order.

Reply:
> `BTreeMap` is sorted by key. `IndexMap` maintains insertion order.

But for log events don't we want to compare using the insertion order? If there are two JSON documents consumed from `stdin` source with different fields order, I would consider them to be different.

> Consider implementing a slower workaround for now - or just switching to `IndexMap` right away.

Initially I implemented sorting before the comparison, but later decided that the actual comparison order should be not lexicographical order, but the insertion order, and removed it (see https://github.com/timberio/vector/pull/1662/commits/9d63e101afa1683704197142f06b137583a9a37f).

But it is an interesting question to discuss, what kind of ordering is more suited here.

Reply:
Is insertion order relevant for a discriminant? I'd say there may be cases where user would want that, however, in the majority of the scenarios they'll only be interested in the shape of the actual value - which means the order of fields is irrelevant, and to maps with the same fields but in different order are considered equal. That would mean explicitly ignoring the order of insertions, and hashing entries more deterministically. `BTreeMap` would do the job.

Reply:
@MOZGIII do we have a test that asserts this behavior? If not, could you open an issue representing that work? You can see how @a-rodin is defining behavior tests in this PR via the `tests/behavior` folder.

Reply:
@binarylogic I'm getting a bit confused with the terminology. By test harness, you mean the `tests/behavior` related stuff in this repo, not the [test harness](https://github.com/timberio/vector-test-harness) in that separate one I'm working on, right?
We don't have a test checking that, and I was wondering how can we add it - I'll definitely create one now.

Reply:
Yes, here in this repo, not the remote test harness.

Reply:
Actually, the test that would check the field order probably belongs to a unit test - it's a more controlled environment that would allow us to reliably trigger the ordering inconsistencies. @binarylogic what do you think?
I'd add another test - a test for merge transform to validate the nested fields merging correctness - at a behavior test though. That's from other comments thread though :smile:

Reply:
Created #1806

Reply:
I kicked in the #1807.
With #1806, haw are we going to do this? I create a PR adding the relevant tests to this branch if that works for everyone.
I also just realized that the base implementation we have currently is iterating over the event keys, so it's also susceptible to the issues I described at https://github.com/timberio/vector/issues/1806#issuecomment-587092360.

Reply:
Why not merge into `master`? The behavior tests should be set up in such a way that they shouldn't require anything from this PR.

Reply:
I'm not sure that behavior tests are a good tool for the job. There are too much layers involved that can hide potential problems, leading to false positive. The discriminant unit tests are way more suitable for a test this delicate.

Reply:
I'll add behavior test too, there's a value of checking that in the global scope thing also work as expected. That said, the core problem is definitely worth addressing and testing explicitly in the units. I didn't realize it was critical until I wrote [this](https://github.com/timberio/vector/issues/1806#issuecomment-587092360).

Reply:
Actually, for the non-nested case the hasing order is well-defined via the order of the `discriminant_fields` slice. False alarm. The `Map` case is still problematic.

Reply:
I created #1825. I can change the PR base to this branch so that we can merge into here, or we can merge that to `master` (doesn't sound like a good idea while this is WIP). Cherry-picking commits here and letting GihHub close that PR when this one is merged is also a viable solution.